_posts/2022-08-04-calldata-tuple-reencoding-head-overflow-bug.md

```diff
@@ -0,0 +1,219 @@
+```
+
+should return the following values:
+```
+0x0000000000000000000000000000000000000000000000000000000000000001 a
+0x0000000000000000000000001111111111222222222233333333334444444444 b.x
+0x000000000000000000000000000000000000000000000000000000000000000b b.y[0]
+0x000000000000000000000000000000000000000000000000000000000000000c b.y[1]
+0x000000000000000000000000000000000000000000000000000000000000000d b.y[2]
+0x6100000000000000000000000000000000000000000000000000000000000000 c[0]
+0x6200000000000000000000000000000000000000000000000000000000000000 c[1]
+```
+
+The encoded tuple has no tail and is laid out in the following way:
+```
+|---------------------------------------------------|
+|                        HEAD                       |
+|------------+-------------------------+------------|
+| value of a |        value of b       | value of c |
+|    bool    |             S           | bytes32[2] |
+|            |-------------------------|            |
+|            |          HEAD of S      |            |
+|            |------------+------------|            |
+|            | value of x | value of y |            |
+|            |   address  |   uint[3]  |            |
+|------------+------------+------------+------------|
+|     1      |      2     |     3      |      4     |
+|------------+------------+------------+------------|
+```
+
+The numbers at the bottom indicate the order in which the components are written by the encoder.
+
+#### Dynamically-sized Example
+
+The example below uses a very similar tuple but with one of the statically-sized fields replaced by a dynamically-sized one:
+```solidity
+struct T {
+    bytes x;
+    uint[3] y;
+}
+
+contract D {
+    function f(bool a, T calldata b, bytes32[2] calldata c)
+        public
+        returns (bool, T calldata, bytes32[2] calldata)
+    {
+        return (a, b, c);
+    }
+}
+```
+
+In this case a call to `D.f()` with the following input:
+```solidity
+f(true, T("abcd", [uint(11), 12, 13]), [bytes32("a"), "b"])
+```
+results in a vastly different encoding:
+```
+0x0000000000000000000000000000000000000000000000000000000000000001 a
+0x0000000000000000000000000000000000000000000000000000000000000080 offset of b
+0x6100000000000000000000000000000000000000000000000000000000000000 c[0]
+0x6200000000000000000000000000000000000000000000000000000000000000 c[1]
+0x0000000000000000000000000000000000000000000000000000000000000080 offset of b.x
+0x000000000000000000000000000000000000000000000000000000000000000b b.y[0]
+0x000000000000000000000000000000000000000000000000000000000000000c b.y[1]
+0x000000000000000000000000000000000000000000000000000000000000000d b.y[2]
+0x0000000000000000000000000000000000000000000000000000000000000004 b.x (length field)
+0x6162636400000000000000000000000000000000000000000000000000000000 b.x (data)
+```
+
+The introduction of a dynamically-sized element resulted in `T` becoming dynamically-sized as well.
```

Comment:
Ah, yeah, this relates to what I meant in https://github.com/ethereum/solidity-blog/pull/187/files#r936962743...
Not sure if it's clearer to think of ``T`` as "becoming dynamically-sized"... it contains a dynamically sized type, but one can still view ``T`` itself as statically-sized... internally we say ```T`` becomes "dynamically-encoded"... but yeah, still not sure if reproducing that distinction is helpful :-).

Reply:
Given the phrasing in https://docs.soliditylang.org/en/v0.8.15/abi-spec.html we could at least change it to
```suggestion
The introduction of a dynamically-sized element resulted in `T` itself becoming a dynamic type.
```
or so...

Reply:
I'm going to stick to "dynamically-sized" as I said in https://github.com/ethereum/solidity-blog/pull/187#discussion_r937504773 but I'll change it to `` `T` itself becoming a dynamically-sized type``.